File: Design and Analysis of Algorithms/karatsuba.cpp
// Convert int to string
#include <algorithm>
#include <cmath>
#include <iostream>
#include <ostream>
#include <string>

std::string to_string(int n) {
  std::string s;

  while (n > 0) {
    s += (n % 10) + '0';
    n /= 10;
  }

  std::reverse(s.begin(), s.end());
  return s;
}

long karatsuba(int num1, int num2) {

  // std ::cout << "Current call is: " << num1 << ", " << num2 << std::endl;

  int n = std::max(to_string(num1).length(), to_string(num2).length());

  if (num1 <= 10 || num2  <= 10) {
    return num1 * num2;
  }


  int nby2 = n / 2;

  int a = num1 / pow(10, nby2);
  int b = num1 % (int)pow(10, nby2);

  // std ::cout << "(a, b): " << a << ", " << b << std::endl;

  int c = num2 / pow(10, nby2);
  int d = num2 % (int)pow(10, nby2);

  // std ::cout << "(c, d): " << c << ", " << d << std::endl;

  long X = karatsuba(a, c);
  long Y = karatsuba(b, d);
  long Z = karatsuba(a + b, c + d) - X - Y;

  long res = X * pow(10, 2 * nby2) + Z * pow(10, nby2) + Y;

  return res;
}

int main() {

  int num1 = 14613;
  int num2 = 35210;

  std :: cout << num1 << " * " << num2 << " = ";

  std::cout <<  karatsuba(num1, num2) << std::endl;

  return 0;
}

File: Advanced Competetive Coding/Karatsuba.java
import java.util.Scanner;

public class Karatsuba {
  private static String toString(int n) {
    StringBuilder s = new StringBuilder();

    while (n > 0) {
      s.append((char) ((n % 10) + '0'));
      n /= 10;
    }

    return s.reverse().toString();
  }

  public static long karatsuba(int num1, int num2) {
    int n = Math.max(toString(num1).length(), toString(num2).length());

    if (num1 <= 10 || num2 <= 10) {
      return (long) num1 * num2;
    }

    int nby2 = n / 2;

    int a = num1 / (int) Math.pow(10, nby2);
    int b = num1 % (int) Math.pow(10, nby2);

    int c = num2 / (int) Math.pow(10, nby2);
    int d = num2 % (int) Math.pow(10, nby2);

    long X = karatsuba(a, c);
    long Y = karatsuba(b, d);
    long Z = karatsuba(a + b, c + d) - X - Y;

    long res = (long) (X * Math.pow(10, 2 * nby2) + Z * Math.pow(10, nby2) + Y);

    return res;
  }

  public static int binaryToDecimal(int num) {
    int decimal = 0;
    int n = 0;

    while (true) {
      if (num == 0) {
        break;
      } else {
        int temp = num % 10;
        decimal += temp * Math.pow(2, n);
        num = num / 10;
        n++;
      }
    }

    return decimal;
  }

  public static void main(String[] args) {

    Scanner sc = new Scanner(System.in);
    int num1b = sc.nextInt();
    int num2b = sc.nextInt();
    sc.close();

    int num1 = binaryToDecimal(num1b);
    int num2 = binaryToDecimal(num2b);

    System.out.println(karatsuba(num1, num2));
  }
}
